src/libraries/Microsoft.PowerFx.Interpreter/Functions/IRandomService.cs

```diff
@@ -38,5 +47,28 @@ public virtual double NextDouble()
                 return _random.NextDouble();
             }
         }
+
+        public virtual decimal NextDecimal()
+        {
+            lock (_randomizerLock)
+            {
+                var randChars = new char[30];
+                randChars[0] = '0';
+                randChars[1] = '.';
+
+                for (int i = 2; i < 30; i++)
+                {
+                    randChars[i] = (char)_random.Next(48, 58);
+                }
+
+                var randStr = new string(randChars);
+                if (decimal.TryParse(randStr, out decimal decimalResult))
+                {
+                    return decimalResult;
+                }
+
+                return decimal.MaxValue;
```

Comment:
>decimal.MaxValue [](http://example.com/codeflow?start=23&length=16)

We shouldn't return decimal.MaxValue - this function should always return something between [0,1). 0.5?

Reply:
Note this means it's not truly a linear distribution:  there is a higher chance of getting .5 than other values.

Reply:
This shouldn't actually happen - the number created above is well-formed, so this is more of a "safeguard". We could, however, change it to `return (decimal)_random.NextDouble();` to be safe

Reply:
There is an interesting discussion on SO here: 
[https://stackoverflow.com/questions/609501/generating-a-random-decimal-in-c-sharp](https://stackoverflow.com/questions/609501/generating-a-random-decimal-in-c-sharp)

Reply:
We started from it :-). As it is stated in the discussion, the distribution in most cases is not uniform, so we iterated through some and came up with that algorithm which works for our scenario (we don't need a random decimal, we just need one between 0-1 that has as much precision as possible. Decimal can have up to 29 digits, but with 29 it would only go to a 7.9 mantissa, so limiting it to 28 significant digits gives us a uniform distribution that other distributions couldn't provide.